test: smoke test for test project manual test and sample scenes

NoelStephensUnity opened the pull request:
## Test Project Smoke Test (Step 1)
Provide an automated "smoke test" for our manual test and sample scenes that:
 * Runs during CI along with all other tests
   * Is a minimal work effort (within reason) to determine that we are not breaking manual tests
* Verifies all registered menu item related scenes are registered in the scenes in build list and can be loaded 
  * Validates that all menu options should work
* Verifies that all primary scenes in each SceneReference can be started as a host and the host can send RPCs to itself
     * Validates that the basic/minimal NetworkManager configuration requirements are met

[MTT-1364](https://jira.unity3d.com/browse/MTT-1364)

## Fixes:
[MTT-1484](https://jira.unity3d.com/browse/MTT-1484):  
Fix for this bug is included and same region of code was required to fix for smoke test.


**This also includes:**
* several test project specific updates based on the recent runtime SDK modifications.
* fixes that were required in order for the smoke test to finish checking all scenes referenced.


## Testing and Documentation

* Smoke Tests Added:
  * SmokeTestStateTest - Validates the smoke test is working properly
  * RegisteredSceneValidation - Validates the menu scene references are in the build settings scenes in build list.
  * SceneReferenceValidation - Validates the primary scene for each scene reference (i.e. contains the NetworkManager) can be loaded, the NetworkManager can be started as a host, and the host can send RPCs to itself to validate the basic configuration needs for the NetworkManager are met.

NoelStephensUnity committed:
update

Just updating to various changes over the past month.

NoelStephensUnity committed:
Merge branch 'develop' into test/smoketest-validation-manual-tests

NoelStephensUnity committed:
Merge branch 'develop' into test/smoketest-validation-manual-tests

NoelStephensUnity committed:
Merge branch 'develop' into test/smoketest-validation-manual-tests

NoelStephensUnity committed:
test and update

Adding smoke test framework and initial registered scene validation test.
Updating some of the test project specific assets including CommandLineHandler which needed a UTP/UnityTransport entry for address and port.  Added two methods to UnityTransport that allowed one to register address and port individually.

NoelStephensUnity committed:
Merge branch 'develop' into test/smoketest-validation-manual-tests

NoelStephensUnity committed:
Merge branch 'develop' into test/smoketest-validation-manual-tests

NoelStephensUnity committed:
refactor

RegisteredSceneSmokeTest now unloads scenes after it has pulled pertinent information from them as opposed to unloading them all at the end.

Renamed erroneously named SmokeTestState,cs to SmokeTest.cs.

Increased default time out of the SmokeTestOrechestrator to 30 seconds.

NoelStephensUnity committed:
refactor and fix

Changed when scenes are unloaded to occur once the scene loaded is no longer needed to reduce scene overhead.  Keeping the scenes list to be tested.
SmokTestOrchestrator is constructed during OneTimeSetup now.

Fixed a bug in the SmokTestOrchestrator that would only allow one state to be invoked.

Migrated all logging of what phase a SmokeTestState is in to the SmokeTestState itself.

NoelStephensUnity committed:
Merge branch 'develop' into test/smoketest-validation-manual-tests

NoelStephensUnity committed:
merge fix

Fixing differences in UTP address and port assignment.

NoelStephensUnity committed:
Merge branch 'develop' into test/smoketest-validation-manual-tests

NoelStephensUnity committed:
Refactor and Fix

Refactored how scenes are grouped in order to determine what might need to do scene transitions and what only has 1 scene associated with it (this might go away in the future).

Re-organized a few things to create the SceneAwareSmokeTestState class that includes common scene loading and unloading methods and checks.

Fixed an issue in the NetworkPrefabPool class where it was throwing an exception (seemingly depending upon how many clients were connected) and that would cause an issue with the client's synchronization process.  Just checking for IsSpawned within DeregisterCustomPrefabHandler fixes this issue.

NoelStephensUnity committed:
test and fix

Fixed the issue with NetworkRigidBody throwing an exception in FixedUpdate (required in order to complete the initial smoke tests).

Fixed issue with the NetworkAnimatorEnhancement scene's NetworkManager not having a valid transport assigned.

Finalized the first pass smoke tests for test project manual tests and samples.

NoelStephensUnity committed:
Merge branch 'develop' into test/smoketest-validation-manual-tests

NoelStephensUnity committed:
Merge branch 'develop' into test/smoketest-validation-manual-tests

NoelStephensUnity commented on com.unity.netcode.gameobjects/Components/NetworkRigidbody.cs:
```diff
@@ -32,7 +32,7 @@ private void Awake()
         private void FixedUpdate()
         {
-            if (NetworkManager.IsListening)
+            if (IsSpawned && NetworkManager.IsListening)
```
This fix is required in order for the smoke test to validate all scenes, otherwise scenes that include NetworkObjects with the NetworkRigidbody component will throw an exception when unloading.

mattwalsh-unity replied:
Interesting.  So in NetworkTransform, we similarly (but not exactly) check for NetworkObject.IsSpawned, but we do not check NM.IsListening.  Seems to me we only need IsSpawned then - how could you be spawned if your NM isn't connected?

Then I notice that NetworkAnimator by way of SendMessagesAllowed is also checking IsSpawned.

So after all that I think my suggestion would be to reduce this just to `IsSpawned`

NoelStephensUnity replied:
Fixed.

NoelStephensUnity commented on com.unity.netcode.adapter.utp/Runtime/UnityTransport.cs:
```diff
@@ -348,6 +348,16 @@ public void SetConnectionData(NetworkEndPoint endPoint)
+        {
+            ConnectionData.Port = port;
+        }
+
```
These two methods were needed to be able to implement command line settings (i.e. you might want to only set one or the other).

mattwalsh-unity replied:
Maybe it's too far for this PR, but what about having an IAddressPort or something which Unet & UTP would have and then these 2 methods wouldn't have to be ad-hoc @andrews-unity?

andrews-unity replied:
So in the transport API overall we will have a unified way to pass down information like this. But these functions aren't actually needed for this use case you can just use: 
`unityTransport.ConnectionData.Address = address;`
`unityTransport.ConnectionData.Port = port;`

NoelStephensUnity replied:
This makes setting each property much easier with no methods required!  Awesome!

NoelStephensUnity commented on testproject/Assets/Scripts/MenuManagement/MainMenuManager.cs:
```diff
@@ -3,12 +3,34 @@
 using UnityEngine.SceneManagement;
 using UnityEngine.UI;
 
+
```
All of these changes are needed to be able to validate Menu and Scene references.

mattwalsh-unity replied:
Can we not just re-use `MainMenuManager` vs. have to make the duplicate `SceneMenuManager` class?

NoelStephensUnity replied:
This code is not part of the PR and will hopefully be replaced with better code in the near future.
I was thinking about migrating my bootstrap code base into test project.
Let me know if you want me to create a PR for this.

NoelStephensUnity commented on testproject/Assets/Scripts/MenuManagement/SceneMenuManager.cs:
```diff
@@ -1,7 +1,29 @@
 /// </summary>
 public class SceneMenuManager : MenuManager<SceneReference>
 {
+    public static List<SceneMenuManager> SceneMenuManagers = new List<SceneMenuManager>();
```
All of these additions are used by the smoke test.

mattwalsh-unity replied:
See above to avoid code duplication

NoelStephensUnity replied:
This is not code duplication.
MainMenuManager handles menu references.
SceneMenuManager handles scene references that are contained within a menu,.

I agree this is redundant, and should be replaced with something better... perhaps the golden path boot strap sample.
Right now I am trying to get the smoke test working with minimal changes.

NoelStephensUnity committed:
style

Adding some additional comments, removing extra LFs, and setting the debug logging to NONE for smoke test.

NoelStephensUnity commented on testproject/Assets/Tests/Runtime/testproject.runtimetests.asmdef:
```diff
@@ -9,7 +9,8 @@
         "Unity.Netcode.Components",
         "Sample",
-        "ClientNetworkTransform"
+        "ClientNetworkTransform",
```
Smoke testing needed a reference back to the TestProject for menu and scene references.

NoelStephensUnity committed:
style

fixed several white space issues.

NoelStephensUnity committed:
fix

more than one of the same command line arguments check.
(still means there could be other related issues or the command line has duplicate arguments)

NoelStephensUnity committed:
Merge branch 'develop' into test/smoketest-validation-manual-tests

NoelStephensUnity committed:
Merge branch 'develop' into test/smoketest-validation-manual-tests

mattwalsh-unity commented on testproject/Assets/Scripts/CommandLineHandler.cs:
```diff
@@ -48,8 +48,10 @@ public CommandLineProcessor(string[] args)
+                if (!m_CommandLineArguments.ContainsKey(arg))
+                {
+                    m_CommandLineArguments.Add(arg, value);
+                }
```
Please put in a `Debug.LogWarning($"Duplicate command line argument for {arg} - ignoring")`

mattwalsh-unity commented on testproject/Assets/Scripts/CommandLineHandler.cs:
```diff
@@ -215,6 +219,9 @@ private void SetPort(ushort port)
                 unetTransport.ConnectPort = port;
                 unetTransport.ServerListenPort = port;
                 break;
+            case UnityTransport unityTransport:
```
Yeah, like I said above I really wish this was hidden behind an interface but I also understand that can be for a later PR.  @zain-mecklai this was the thing you were looking for, yes?

zain-mecklai replied:
Yes!

NoelStephensUnity replied:
The relatively new UnityTransport.ConnectionData lets you assign directly.  This removes the need for those methods.

mattwalsh-unity commented on testproject/Assets/Tests/Manual/Scripts/GenericNetworkObjectBehaviour.cs:
```diff
@@ -102,7 +102,7 @@ public void SetDirectionAndVelocity(Vector3 direction, float velocity)
         private void FixedUpdate()
         {
-            if (NetworkManager != null && NetworkManager.IsListening)
+            if (IsSpawned && NetworkManager != null && NetworkManager.IsListening)
```
See above, should not need `NetworkManager.IsListening`, people will copy & paste

mattwalsh-unity commented on testproject/Assets/Tests/Manual/Scripts/GenericNetworkObjectBehaviour.cs:
```diff
@@ -140,27 +140,38 @@ private void FixedUpdate()
                     }
                 }
             }
+            else if (!IsSpawned && IsRegisteredPoolObject)
```
This can be just `else if (IsRegisteredPoolObject)`

mattwalsh-unity commented on testproject/Assets/Tests/Manual/Scripts/GenericNetworkObjectBehaviour.cs:
```diff
@@ -140,27 +140,38 @@ private void FixedUpdate()
             }
+            else if (!IsSpawned && IsRegisteredPoolObject)
+            {
+                if (m_MeshRenderer != null && m_MeshRenderer.enabled)
```
I believe this can be `if (m_MeshRenderer?.enabled`

andrews-unity replied:
No :P we can't use the NullCoalesce operator :P I just made a post about this is slack.

mattwalsh-unity commented on testproject/Assets/Tests/Manual/Scripts/NetworkPrefabPool.cs:
```diff
@@ -165,7 +165,7 @@ private void RegisterCustomPrefabHandler()
         {
             // Register the custom spawn handler?
-            if (EnableHandler && NetworkManager && NetworkManager.PrefabHandler != null)
+            if (EnableHandler && IsSpawned && NetworkManager && NetworkManager.PrefabHandler != null)
```
Is it possible for `IsSpawned` to be true but `NetworkManager && NetworkManager.PrefabHandler != null` to be false?  Don't think so, do the simplest versoin

mattwalsh-unity reviewed (changes requested):
A number of small-ish things.  Also I'm not exactly sure were to start in the 300+ lines of new test project code which I assume is the "run through the tests" thing

NoelStephensUnity committed:
refactor

Refactors some code that could be further simplified.

NoelStephensUnity committed:
refactor and style

refactoring to simplest logic.
removing whitespace.

NoelStephensUnity committed:
Merge branch 'develop' into test/smoketest-validation-manual-tests

NoelStephensUnity committed:
update

adding debug information about duplicate command line arguments.

NoelStephensUnity committed:
update

Adding some of the minor updates during manual testing release branches.

NoelStephensUnity committed:
Merge branch 'develop' into test/smoketest-validation-manual-tests

NoelStephensUnity committed:
revert

reverting  change from OnDestroy to OnNetworkDespawn in NetworkPrefabPoolAdditive.

jeffreyrainy commented on com.unity.netcode.gameobjects/Components/NetworkRigidbody.cs:
```diff
@@ -32,13 +32,10 @@ private void Awake()
         private void FixedUpdate()
         {
-            if (NetworkManager.IsListening)
+            if (IsSpawned && (HasAuthority != m_IsAuthority))
```
Curiosity, why/how does NetworkRigidBody gets modified if it is a test ?

jeffreyrainy replied:
nvm, just saw the conversation on the, well..., conversation tab :-)

NoelStephensUnity replied:
There are some updates in this PR that fixed outstanding issues in some areas that would throw exceptions under specific scenarios (i.e. shutting down or the like).  The above change assures that NetworkRigidBody doesn't throw an exception upon shutting down the NetworkManager.  
Since the Smoke Tests will run during CI, it requires that there are no exceptions while performing the smoke tests.
